File: node/tests/result.test.ts
/*
 * Copyright 2024 ScopeDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import assert from "node:assert/strict";
import { describe, it } from "node:test";
import type { DataType } from "../src/protocol.js";
import { FieldSchema, ResultSet, Schema } from "../src/result.js";
import { ScopeDBError } from "../src/errors.js";

function makeResultSet(
  fields: Array<{ name: string; data_type: DataType }>,
  rows: Array<Array<string | null>>,
): ResultSet {
  return ResultSet.fromStatementResultSet({
    metadata: { fields, num_rows: rows.length },
    format: "json",
    rows,
  });
}

describe("ResultSet.fromStatementResultSet", () => {
  it("builds schema from field metadata", () => {
    const rs = makeResultSet(
      [{ name: "id", data_type: "int" }, { name: "name", data_type: "string" }],
      [],
    );
    const fields = rs.schema().fields();
    assert.equal(fields.length, 2);
    assert.equal(fields[0]!.name(), "id");
    assert.equal(fields[0]!.dataType(), "int");
    assert.equal(fields[1]!.name(), "name");
    assert.equal(fields[1]!.dataType(), "string");
  });

  it("reports correct numRows", () => {
    const rs = makeResultSet(
      [{ name: "x", data_type: "int" }],
      [["1"], ["2"], ["3"]],
    );
    assert.equal(rs.numRows(), 3);
  });

  it("preserves raw JSON rows via jsonRows()", () => {
    const rs = makeResultSet(
      [{ name: "x", data_type: "int" }, { name: "y", data_type: "string" }],
      [["1", "hello"], ["2", null]],
    );
    const rows = rs.jsonRows();
    assert.deepEqual(rows[0], ["1", "hello"]);
    assert.deepEqual(rows[1], ["2", null]);
  });
});

describe("ResultSet.intoValues — type parsing", () => {
  it("parses int as BigInt", () => {
    const rs = makeResultSet([{ name: "n", data_type: "int" }], [["42"]]);
    assert.equal(rs.intoValues()[0]![0], 42n);
  });

  it("parses uint as BigInt", () => {
    const rs = makeResultSet([{ name: "n", data_type: "uint" }], [["100"]]);
    assert.equal(rs.intoValues()[0]![0], 100n);
  });

  it("parses u_int (backward-compat alias) as BigInt", () => {
    const rs = makeResultSet([{ name: "n", data_type: "u_int" }], [["99"]]);
    assert.equal(rs.intoValues()[0]![0], 99n);
  });

  it("parses float as number", () => {
    const rs = makeResultSet([{ name: "f", data_type: "float" }], [["3.14"]]);
    const val = rs.intoValues()[0]![0];
    assert.ok(typeof val === "number");
    assert.ok(Math.abs((val as number) - 3.14) < 1e-9);
  });

  it("parses boolean true", () => {
    const rs = makeResultSet([{ name: "b", data_type: "boolean" }], [["true"]]);
    assert.equal(rs.intoValues()[0]![0], true);
  });

  it("parses boolean false", () => {
    const rs = makeResultSet([{ name: "b", data_type: "boolean" }], [["false"]]);
    assert.equal(rs.intoValues()[0]![0], false);
  });

  it("parses timestamp as Date", () => {
    const rs = makeResultSet([{ name: "t", data_type: "timestamp" }], [["2024-06-01T12:00:00Z"]]);
    const val = rs.intoValues()[0]![0];
    assert.ok(val instanceof Date);
    assert.equal((val as Date).toISOString(), "2024-06-01T12:00:00.000Z");
  });

  it("returns string as-is for string type", () => {
    const rs = makeResultSet([{ name: "s", data_type: "string" }], [["hello"]]);
    assert.equal(rs.intoValues()[0]![0], "hello");
  });

  it("returns string as-is for interval type", () => {
    const rs = makeResultSet([{ name: "d", data_type: "interval" }], [["P1D"]]);
    assert.equal(rs.intoValues()[0]![0], "P1D");
  });

  it("returns null for null cell regardless of type", () => {
    const types: DataType[] = ["int", "uint", "float", "boolean", "timestamp", "string"];
    for (const dt of types) {
      const rs = makeResultSet([{ name: "v", data_type: dt }], [[null]]);
      assert.equal(rs.intoValues()[0]![0], null, `expected null for type ${dt}`);
    }
  });

  it("throws ScopeDBError on invalid integer value", () => {
    const rs = makeResultSet([{ name: "n", data_type: "int" }], [["not-a-number"]]);
    assert.throws(() => rs.intoValues(), ScopeDBError);
  });

  it("throws ScopeDBError on invalid float value", () => {
    const rs = makeResultSet([{ name: "f", data_type: "float" }], [["NaN"]]);
    assert.throws(() => rs.intoValues(), ScopeDBError);
  });

  it("throws ScopeDBError on invalid boolean value", () => {
    const rs = makeResultSet([{ name: "b", data_type: "boolean" }], [["yes"]]);
    assert.throws(() => rs.intoValues(), ScopeDBError);
  });

  it("throws ScopeDBError on invalid timestamp value", () => {
    const rs = makeResultSet([{ name: "t", data_type: "timestamp" }], [["not-a-date"]]);
    assert.throws(() => rs.intoValues(), ScopeDBError);
  });

  it("throws ScopeDBError when row field count mismatches schema", () => {
    const rs = makeResultSet(
      [{ name: "a", data_type: "string" }, { name: "b", data_type: "string" }],
      [["only-one-cell"]],
    );
    assert.throws(() => rs.intoValues(), ScopeDBError);
  });

  it("handles multiple rows with mixed types correctly", () => {
    const rs = makeResultSet(
      [
        { name: "id", data_type: "int" },
        { name: "score", data_type: "float" },
        { name: "label", data_type: "string" },
        { name: "active", data_type: "boolean" },
      ],
      [
        ["1", "9.5", "alpha", "true"],
        ["2", "3.0", "beta", "false"],
      ],
    );
    const rows = rs.intoValues();
    assert.equal(rows[0]![0], 1n);
    assert.equal(rows[1]![2], "beta");
    assert.equal(rows[1]![3], false);
  });
});

describe("Schema and FieldSchema", () => {
  it("returns fields in order", () => {
    const schema = new Schema([
      new FieldSchema("a", "int"),
      new FieldSchema("b", "string"),
      new FieldSchema("c", "boolean"),
    ]);
    const fields = schema.fields();
    assert.equal(fields.length, 3);
    assert.equal(fields[2]!.name(), "c");
    assert.equal(fields[2]!.dataType(), "boolean");
  });
});

File: go/result_test.go
/*
 * Copyright 2024 ScopeDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package scopedb

import (
	"encoding/json"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
)

func TestResultSetToValuesConvertsBinaryAndInterval(t *testing.T) {
	t.Parallel()

	rows, err := json.Marshal([][]*string{{
		stringPointer("0001ABFF"),
		stringPointer(""),
		stringPointer("abcdef"),
		stringPointer("PT1H2M3.000000004S"),
		stringPointer("-PT1M0.000000001S"),
		nil,
	}})
	require.NoError(t, err)

	resultSet := &ResultSet{
		TotalRows: 1,
		Schema: Schema{
			&FieldSchema{Name: "payload", Type: BinaryDataType},
			&FieldSchema{Name: "empty_payload", Type: BinaryDataType},
			&FieldSchema{Name: "lowercase_payload", Type: BinaryDataType},
			&FieldSchema{Name: "elapsed", Type: IntervalDataType},
			&FieldSchema{Name: "offset", Type: IntervalDataType},
			&FieldSchema{Name: "nothing", Type: NullDataType},
		},
		rows: rows,
	}

	values, err := resultSet.ToValues()
	require.NoError(t, err)
	require.Equal(t, [][]Value{{
		[]byte{0x00, 0x01, 0xab, 0xff},
		[]byte{},
		[]byte{0xab, 0xcd, 0xef},
		time.Hour + 2*time.Minute + 3*time.Second + 4*time.Nanosecond,
		-time.Minute - time.Nanosecond,
		nil,
	}}, values)
}

func TestResultSetRawRowsReturnsWireCells(t *testing.T) {
	t.Parallel()

	resultSet := &ResultSet{
		TotalRows: 2,
		Schema: Schema{
			&FieldSchema{Name: "value", Type: StringDataType},
			&FieldSchema{Name: "optional", Type: StringDataType},
		},
		rows: json.RawMessage(`[["first",null],["second","present"]]`),
	}

	rows, err := resultSet.RawRows()
	require.NoError(t, err)
	require.Len(t, rows, 2)
	require.Equal(t, "first", *rows[0][0])
	require.Nil(t, rows[0][1])
	require.Equal(t, "second", *rows[1][0])
	require.Equal(t, "present", *rows[1][1])
}

func TestResultSetToObjectsKeysConvertedValues(t *testing.T) {
	t.Parallel()

	resultSet := &ResultSet{
		TotalRows: 1,
		Schema: Schema{
			&FieldSchema{Name: "signed", Type: IntDataType},
			&FieldSchema{Name: "unsigned", Type: UIntDataType},
			&FieldSchema{Name: "at", Type: TimestampDataType},
			&FieldSchema{Name: "elapsed", Type: IntervalDataType},
			&FieldSchema{Name: "payload", Type: BinaryDataType},
		},
		rows: json.RawMessage(`[[
			"-7",
			"9",
			"2026-08-08T00:00:00.123456789Z",
			"PT1.000000002S",
			"00ff"
		]]`),
	}

	objects, err := resultSet.ToObjects()
	require.NoError(t, err)
	require.Equal(t, []map[string]Value{{
		"signed":   int64(-7),
		"unsigned": uint64(9),
		"at":       time.Date(2026, 8, 8, 0, 0, 0, 123456789, time.UTC),
		"elapsed":  time.Second + 2*time.Nanosecond,
		"payload":  []byte{0x00, 0xff},
	}}, objects)
}

func TestResultSetFirstOnlyConvertsFirstRow(t *testing.T) {
	t.Parallel()

	resultSet := &ResultSet{
		TotalRows: 2,
		Schema:    Schema{&FieldSchema{Name: "value", Type: IntDataType}},
		rows:      json.RawMessage(`[["1"],["not-an-integer"]]`),
	}

	first, ok, err := resultSet.First()
	require.NoError(t, err)
	require.True(t, ok)
	require.Equal(t, map[string]Value{"value": int64(1)}, first)

	_, err = resultSet.ToObjects()
	require.Error(t, err, "converting every row must still surface the invalid second row")
}

func TestResultSetFirstReportsEmptyResult(t *testing.T) {
	t.Parallel()

	resultSet := &ResultSet{
		Schema: Schema{&FieldSchema{Name: "value", Type: StringDataType}},
		rows:   json.RawMessage(`[]`),
	}

	first, ok, err := resultSet.First()
	require.NoError(t, err)
	require.False(t, ok)
	require.Nil(t, first)
}

func TestResultSetObjectConversionsRejectDuplicateColumns(t *testing.T) {
	t.Parallel()

	resultSet := &ResultSet{
		TotalRows: 1,
		Schema: Schema{
			&FieldSchema{Name: "value", Type: IntDataType},
			&FieldSchema{Name: "value", Type: IntDataType},
		},
		rows: json.RawMessage(`[["1","2"]]`),
	}

	_, err := resultSet.ToObjects()
	require.ErrorContains(t, err, `duplicate result column name "value"`)
	_, _, err = resultSet.First()
	require.ErrorContains(t, err, `duplicate result column name "value"`)
}

func TestResultSetToValuesConvertsIntervalBoundaries(t *testing.T) {
	t.Parallel()

	tests := map[string]time.Duration{
		"PT0S":                        0,
		"PT24H":                       24 * time.Hour,
		"PT2562047H47M16.854775807S":  time.Duration(1<<63 - 1),
		"-PT2562047H47M16.854775808S": time.Duration(-1 << 63),
	}
	for value, expected := range tests {
		value, expected := value, expected
		t.Run(value, func(t *testing.T) {
			t.Parallel()

			resultSet := resultSetWithSingleValue(IntervalDataType, value)
			values, err := resultSet.ToValues()
			require.NoError(t, err)
			require.Equal(t, expected, values[0][0])
		})
	}
}

func TestResultSetToValuesRejectsInvalidBinary(t *testing.T) {
	t.Parallel()

	resultSet := resultSetWithSingleValue(BinaryDataType, "not-hex")

	_, err := resultSet.ToValues()
	require.ErrorContains(t, err, `invalid binary value "not-hex"`)
}

func TestResultSetToValuesRejectsInvalidInterval(t *testing.T) {
	t.Parallel()

	for _, value := range []string{"1h", "P1D", "PT", "PT1.0000000000S", "PT2562048H"} {
		value := value
		t.Run(value, func(t *testing.T) {
			t.Parallel()

			resultSet := resultSetWithSingleValue(IntervalDataType, value)
			_, err := resultSet.ToValues()
			require.ErrorContains(t, err, "invalid interval value")
		})
	}
}

func TestResultSetToValuesRejectsNonNullValueForNullType(t *testing.T) {
	t.Parallel()

	resultSet := resultSetWithSingleValue(NullDataType, "null")

	_, err := resultSet.ToValues()
	require.ErrorContains(t, err, "unexpected non-null value for null data type")
}

func TestDataTypeUnmarshalAcceptsCanonicalUIntAndHistoricalAlias(t *testing.T) {
	t.Parallel()

	for _, wireValue := range []string{`"uint"`, `"u_int"`} {
		var dataType DataType
		require.NoError(t, json.Unmarshal([]byte(wireValue), &dataType))
		require.Equal(t, UIntDataType, dataType)
	}
}

func TestDataTypeUnmarshalRejectsUnknownAndNonStringValues(t *testing.T) {
	t.Parallel()

	for _, wireValue := range []string{`"unsigned"`, `42`} {
		var dataType DataType
		require.Error(t, json.Unmarshal([]byte(wireValue), &dataType))
	}
}

func resultSetWithSingleValue(dataType DataType, value string) *ResultSet {
	rows, err := json.Marshal([][]*string{{stringPointer(value)}})
	if err != nil {
		panic(err)
	}
	return &ResultSet{
		TotalRows: 1,
		Schema:    Schema{&FieldSchema{Name: "value", Type: dataType}},
		rows:      rows,
	}
}

func stringPointer(value string) *string {
	return &value
}
